CI tests on Mumbai can run out of fake MATIC

### Summary

test_mumbai.py has two tests. Each relies on fake MATIC being in two pre-specified accounts. We're good for local manually-initiated tests. But CI can run out of MATIC, in turn making the test fail. Example traceback below. This is bad, because unit tests should pass, not fail to to external factors like this.

To fix, test_mumbai can auto-grab more fake MATIC if needed.

Details below.

### On local manually-initated tests

developers.md [tells developers](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/developers.md#3-set-up-contracts) how to get that MATIC, as follows.

Some tests run on Mumbai (e.g. test_mumbai.py), which need fake MATIC. So you also need:

#set private keys of two remote accounts
```text
export REMOTE_TEST_PRIVATE_KEY1=<your remote private key 1>
export REMOTE_TEST_PRIVATE_KEY2=<your remote private key 2>
```

These keys aren't public because bots could eat the fake MATIC. You need to generate your own, and fill them with a faucet; [here's how](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/get-test-MATIC.md). (Or, [access-protected OPF keys](https://github.com/oceanprotocol/private-keys/blob/main/README.md)).


### How to fix in CI

To fix, test_mumbai can auto-grab more fake MATIC if needed.

Specifically: when test_mumbai runs, for each account, it can check to see if there's enough MATIC the account. If not, it can query a faucet API to get more. 

Here's an article that uses a faucet api, and how to use it in code. https://medium.com/the-blockchain-presence-blog/getting-more-matic-on-polygon-mumbai-5db25d13480d

If that doesn't work, we can explore other options. This search may help: https://www.google.com/search?q=mumbai+faucet+api


### Example traceback
```python
>>> asset = ocean.assets.create_url_asset(name, url, alice_wallet)

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/enforce_typing/decorator.py", line 29, in wrapper
    return func(*args, **kwargs)
  File "/home/trentmc/code/ocean.py/ocean_lib/ocean/ocean_assets.py", line 266, in create_url_asset
    return self._create1(name, files, publisher_wallet)
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/enforce_typing/decorator.py", line 29, in wrapper
    return func(*args, **kwargs)
  File "/home/trentmc/code/ocean.py/ocean_lib/ocean/ocean_assets.py", line 305, in _create1
    asset = self.create(
  File "/home/trentmc/code/ocean.py/ocean_lib/ocean/ocean_assets.py", line 410, in create
    tx_id = data_nft_factory.deploy_erc721_contract(
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/enforce_typing/decorator.py", line 29, in wrapper
    return func(*args, **kwargs)
  File "/home/trentmc/code/ocean.py/ocean_lib/models/data_nft_factory.py", line 86, in deploy_erc721_contract
    return self.send_transaction(
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/enforce_typing/decorator.py", line 29, in wrapper
    return func(*args, **kwargs)
  File "/home/trentmc/code/ocean.py/ocean_lib/web3_internal/contract_base.py", line 201, in send_transaction
    return contract_function.transact(
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/enforce_typing/decorator.py", line 29, in wrapper
    return func(*args, **kwargs)
  File "/home/trentmc/code/ocean.py/ocean_lib/web3_internal/web3_overrides/contract.py", line 74, in transact
    return transact_with_contract_function(
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/enforce_typing/decorator.py", line 29, in wrapper
    return func(*args, **kwargs)
  File "/home/trentmc/code/ocean.py/ocean_lib/web3_internal/web3_overrides/contract.py", line 137, in transact_with_contract_function
    txn_hash = web3.eth.send_raw_transaction(raw_tx)
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/web3/eth.py", line 818, in send_raw_transaction
    return self._send_raw_transaction(transaction)
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/web3/module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/web3/manager.py", line 198, in request_blocking
    return self.formatted_response(response,
  File "/home/trentmc/code/ocean.py/venv/lib/python3.10/site-packages/web3/manager.py", line 171, in formatted_response
    raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'insufficient funds for gas * price + value'}
```

Another error I would get in Polygon is:
```python
    @staticmethod
    def formatted_response(
        response: RPCResponse,
        params: Any,
        error_formatters: Optional[Callable[..., Any]] = None,
        null_result_formatters: Optional[Callable[..., Any]] = None,
    ) -> Any:
        if "error" in response:
            apply_error_formatters(error_formatters, response)
>           raise ValueError(response["error"])
E           ValueError: {'code': -32000, 'message': 'replacement transaction underpriced'}
```

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -41,13 +41,13 @@ def test_nonocean_tx(tmp_path):
     signed_tx = web3.eth.account.sign_transaction(tx, alice_wallet.private_key)
 
     print("Do a send-Ether tx...")
-    try: #it can get away with "insufficient funds" errors, but not others
+    try:  # it can get away with "insufficient funds" errors, but not others
         tx_hash = web3.eth.send_raw_transaction(signed_tx.rawTransaction)
     except ValueError as error:
         if "insufficient funds" in str(error):
-            warnings.warn(UserWarning("Warning: Insufficient funds (MATIC) in Mumbai testnet"))
+            warnings.warn(UserWarning("Warning: Insufficient test MATIC"))
             return
-        raise(error)
+        raise (error)
 
     print("Wait for send-Ether tx to complete...")
     _ = web3.eth.wait_for_transaction_receipt(tx_hash)
@@ -56,7 +56,7 @@ def test_nonocean_tx(tmp_path):
     assert bob_eth_after > bob_eth_before
 
 
-#@pytest.mark.skip(reason="Don't skip, once fixed #943")
+# @pytest.mark.skip(reason="Don't skip, once fixed #943")
 def test_ocean_tx(tmp_path):
     """Do a (simple) Ocean tx on Mumbai"""
 
@@ -67,14 +67,14 @@ def test_ocean_tx(tmp_path):
 
     # Alice publish data NFT
     print("Do an Ocean tx, and wait for it to complete...")
-    try:  #it can get away with "insufficient funds" errors, but not others
+    try:  # it can get away with "insufficient funds" errors, but not others
         data_nft = ocean.create_data_nft("My NFT1", "NFT1", alice_wallet)
     except ValueError as error:
         if "insufficient funds" in str(error):
-            warnings.warn(UserWarning("Warning: Insufficient funds (MATIC) in Mumbai testnet"))
+            warnings.warn(UserWarning("Warning: Insufficient test MATIC"))
             return
-        raise(error)
-        
+        raise (error)
+
     assert data_nft.symbol() == "NFT1"
 
 

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 import os
+from random import random
 import warnings
 
 import pytest
@@ -23,16 +24,13 @@ def test_nonocean_tx(tmp_path):
 
     # Get gas price (in Gwei) from Polygon gas station
     gas_station_url = "https://gasstation-mumbai.matic.today/v2"
-    base_gas_price = requests.get(gas_station_url).json()["fast"]["maxFee"]
+    gas_price = requests.get(gas_station_url).json()["fast"]["maxFee"]
 
     # Simplest possible tx: Alice send Bob some fake MATIC
     web3 = ocean.web3
     bob_eth_before = web3.eth.get_balance(bob_wallet.address)
     nonce = web3.eth.getTransactionCount(alice_wallet.address)
-
-    # avoid 'replacement transaction underpriced' error: each try, pay more gas
-    gas_price = base_gas_price * (1.0 + nonce / 10.0)
-
+    amt_send = 1e-8 * random()  # fix "replacement tx underpriced": make tx difft
     tx = {
         "nonce": nonce,
         "gasPrice": web3.toWei(gas_price, "gwei"),
@@ -40,7 +38,7 @@ def test_nonocean_tx(tmp_path):
         "chainId": web3.eth.chain_id,
         "to": bob_wallet.address,
         "from": alice_wallet.address,
-        "value": web3.toWei(1e-8, "ether"),
+        "value": web3.toWei(amt_send, "ether"),
     }
     signed_tx = web3.eth.account.sign_transaction(tx, alice_wallet.private_key)
 

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -23,12 +23,16 @@ def test_nonocean_tx(tmp_path):
 
     # Get gas price (in Gwei) from Polygon gas station
     gas_station_url = "https://gasstation-mumbai.matic.today/v2"
-    gas_price = requests.get(gas_station_url).json()["fast"]["maxFee"]
+    base_gas_price = requests.get(gas_station_url).json()["fast"]["maxFee"]
 
     # Simplest possible tx: Alice send Bob some fake MATIC
     web3 = ocean.web3
     bob_eth_before = web3.eth.get_balance(bob_wallet.address)
     nonce = web3.eth.getTransactionCount(alice_wallet.address)
+
+    # avoid 'replacement transaction underpriced' error: each try, pay more gas
+    gas_price = base_gas_price * (1.0 + nonce / 10.0)
+
     tx = {
         "nonce": nonce,
         "gasPrice": web3.toWei(gas_price, "gwei"),

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -56,7 +56,7 @@ def test_nonocean_tx(tmp_path):
     assert bob_eth_after > bob_eth_before
 
 
-# @pytest.mark.skip(reason="Don't skip, once fixed #943")
+@pytest.mark.skip(reason="Don't skip, once fixed #943")
 def test_ocean_tx(tmp_path):
     """Do a (simple) Ocean tx on Mumbai"""
 

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -3,7 +3,8 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 import os
-from random import random
+import random
+import string
 import warnings
 
 import pytest
@@ -30,7 +31,9 @@ def test_nonocean_tx(tmp_path):
     web3 = ocean.web3
     bob_eth_before = web3.eth.get_balance(bob_wallet.address)
     nonce = web3.eth.getTransactionCount(alice_wallet.address)
-    amt_send = 1e-8 * random()  # fix "replacement tx underpriced": make tx difft
+
+    # avoid "replacement transaction underpriced" error: make each tx diff't
+    amt_send = 1e-8 * random.random()
     tx = {
         "nonce": nonce,
         "gasPrice": web3.toWei(gas_price, "gwei"),
@@ -68,16 +71,19 @@ def test_ocean_tx(tmp_path):
     (alice_wallet, _) = _get_wallets(ocean)
 
     # Alice publish data NFT
-    print("Do an Ocean tx, and wait for it to complete...")
+    # avoid "replacement transaction underpriced" error: make each tx diff't
+    cand_chars = string.ascii_uppercase + string.digits
+    symbol = ''.join(random.choices(cand_chars, k=8))
     try:  # it can get away with "insufficient funds" errors, but not others
-        data_nft = ocean.create_data_nft("My NFT1", "NFT1", alice_wallet)
+        print("Do an Ocean tx, and wait for it to complete...")
+        data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)
     except ValueError as error:
         if "insufficient funds" in str(error):
             warnings.warn(UserWarning("Warning: Insufficient test MATIC"))
             return
         raise (error)
 
-    assert data_nft.symbol() == "NFT1"
+    assert data_nft.symbol() == symbol
 
 
 def _get_wallets(ocean):

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -36,7 +36,7 @@ def test_nonocean_tx(tmp_path):
         "chainId": web3.eth.chain_id,
         "to": bob_wallet.address,
         "from": alice_wallet.address,
-        "value": web3.toWei(0.001, "ether"),
+        "value": web3.toWei(1e-8, "ether"),
     }
     signed_tx = web3.eth.account.sign_transaction(tx, alice_wallet.private_key)
 

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -73,7 +73,7 @@ def test_ocean_tx(tmp_path):
     # Alice publish data NFT
     # avoid "replacement transaction underpriced" error: make each tx diff't
     cand_chars = string.ascii_uppercase + string.digits
-    symbol = ''.join(random.choices(cand_chars, k=8))
+    symbol = "".join(random.choices(cand_chars, k=8))
     try:  # it can get away with "insufficient funds" errors, but not others
         print("Do an Ocean tx, and wait for it to complete...")
         data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 import os
+import warnings
 
 import pytest
 import requests
@@ -40,7 +41,13 @@ def test_nonocean_tx(tmp_path):
     signed_tx = web3.eth.account.sign_transaction(tx, alice_wallet.private_key)
 
     print("Do a send-Ether tx...")
-    tx_hash = web3.eth.send_raw_transaction(signed_tx.rawTransaction)
+    try: #it can get away with "insufficient funds" errors, but not others
+        tx_hash = web3.eth.send_raw_transaction(signed_tx.rawTransaction)
+    except ValueError as error:
+        if "insufficient funds" in str(error):
+            warnings.warn(UserWarning("Warning: Insufficient funds (MATIC) in Mumbai testnet"))
+            return
+        raise(error)
 
     print("Wait for send-Ether tx to complete...")
     _ = web3.eth.wait_for_transaction_receipt(tx_hash)
@@ -49,7 +56,7 @@ def test_nonocean_tx(tmp_path):
     assert bob_eth_after > bob_eth_before
 
 
-@pytest.mark.skip(reason="Don't skip, once fixed #943")
+#@pytest.mark.skip(reason="Don't skip, once fixed #943")
 def test_ocean_tx(tmp_path):
     """Do a (simple) Ocean tx on Mumbai"""
 
@@ -60,7 +67,14 @@ def test_ocean_tx(tmp_path):
 
     # Alice publish data NFT
     print("Do an Ocean tx, and wait for it to complete...")
-    data_nft = ocean.create_data_nft("My NFT1", "NFT1", alice_wallet)
+    try:  #it can get away with "insufficient funds" errors, but not others
+        data_nft = ocean.create_data_nft("My NFT1", "NFT1", alice_wallet)
+    except ValueError as error:
+        if "insufficient funds" in str(error):
+            warnings.warn(UserWarning("Warning: Insufficient funds (MATIC) in Mumbai testnet"))
+            return
+        raise(error)
+        
     assert data_nft.symbol() == "NFT1"
 
 

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -72,12 +72,8 @@ def _get_wallets(ocean):
     bob_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY2")
 
     instrs = "You must set it. It must hold Mumbai MATIC."
-    assert (
-        alice_private_key is not None
-    ), f"Need envvar REMOTE_TEST_PRIVATE_KEY1. {instrs}"
-    assert (
-        bob_private_key is not None
-    ), f"Need envvar REMOTE_TEST_PRIVATE_KEY2. {instrs}"
+    assert alice_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY1. {instrs}"
+    assert bob_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY2. {instrs}"
 
     # wallets
     alice_wallet = Wallet(

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -65,26 +65,24 @@ def test_ocean_tx(tmp_path):
 
 
 def _get_wallets(ocean):
-    """Returns (alice_wallet, bob_wallet)"""
+    config, web3 = ocean.config_dict, ocean.web3
 
-    names = ["alice", "bob"]
-    envvars = ["REMOTE_TEST_PRIVATE_KEY1", "REMOTE_TEST_PRIVATE_KEY2"]
-    wallets = [] #fill this
-    for (name, envvar) in zip(names, envvars):
-        private_key = os.getenv(envvar)
+    alice_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY1")
+    bob_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY2")
 
-        instrs = "You must set it. It must hold Mumbai MATIC."
-        assert private_key, f"Need envvar {envvar}. {instrs}"
+    instrs = "You must set it. It must hold Mumbai MATIC."
+    assert alice_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY1. {instrs}"
+    assert bob_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY2. {instrs}"
 
-        # wallets
-        config, web3 = ocean.config_dict, ocean.web3
-        n_confirm, timeout = config["BLOCK_CONFIRMATIONS"], config["TRANSACTION_TIMEOUT"]
-        wallet = Wallet(web3, private_key, n_confirm, timeout)
+    # wallets
+    n_confirm, timeout = config["BLOCK_CONFIRMATIONS"], config["TRANSACTION_TIMEOUT"]
+    alice_wallet = Wallet(web3, alice_private_key, n_confirm, timeout)
+    bob_wallet = Wallet(web3, bob_private_key, n_confirm, timeout)
 
-        print(f"{name}_wallet.address = '{wallet.address}'")
-        wallets.append(wallet)
+    print(f"alice_wallet.address = '{alice_wallet.address}'")
+    print(f"bob_wallet.address = '{bob_wallet.address}'")
 
-    return tuple(wallets)
+    return (alice_wallet, bob_wallet)
 
 
 def _remote_config(tmp_path):

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -76,18 +76,10 @@ def _get_wallets(ocean):
     assert bob_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY2. {instrs}"
 
     # wallets
-    alice_wallet = Wallet(
-        web3,
-        alice_private_key,
-        config["BLOCK_CONFIRMATIONS"],
-        config["TRANSACTION_TIMEOUT"],
-    )
-    bob_wallet = Wallet(
-        web3,
-        bob_private_key,
-        config["BLOCK_CONFIRMATIONS"],
-        config["TRANSACTION_TIMEOUT"],
-    )
+    n_confirm, timeout = config["BLOCK_CONFIRMATIONS"], config["TRANSACTION_TIMEOUT"]
+    alice_wallet = Wallet(web3, alice_private_key, n_confirm, timeout)
+    alice_wallet = Wallet(web3, bob_private_key, n_confirm, timeout)
+
     print(f"alice_wallet.address = '{alice_wallet.address}'")
     print(f"bob_wallet.address = '{bob_wallet.address}'")
 

--- a/tests/integration/remote/test_mumbai.py
+++ b/tests/integration/remote/test_mumbai.py
@@ -65,24 +65,26 @@ def test_ocean_tx(tmp_path):
 
 
 def _get_wallets(ocean):
-    config, web3 = ocean.config_dict, ocean.web3
+    """Returns (alice_wallet, bob_wallet)"""
 
-    alice_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY1")
-    bob_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY2")
+    names = ["alice", "bob"]
+    envvars = ["REMOTE_TEST_PRIVATE_KEY1", "REMOTE_TEST_PRIVATE_KEY2"]
+    wallets = [] #fill this
+    for (name, envvar) in zip(names, envvars):
+        private_key = os.getenv(envvar)
 
-    instrs = "You must set it. It must hold Mumbai MATIC."
-    assert alice_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY1. {instrs}"
-    assert bob_private_key, f"Need envvar REMOTE_TEST_PRIVATE_KEY2. {instrs}"
+        instrs = "You must set it. It must hold Mumbai MATIC."
+        assert private_key, f"Need envvar {envvar}. {instrs}"
 
-    # wallets
-    n_confirm, timeout = config["BLOCK_CONFIRMATIONS"], config["TRANSACTION_TIMEOUT"]
-    alice_wallet = Wallet(web3, alice_private_key, n_confirm, timeout)
-    bob_wallet = Wallet(web3, bob_private_key, n_confirm, timeout)
+        # wallets
+        config, web3 = ocean.config_dict, ocean.web3
+        n_confirm, timeout = config["BLOCK_CONFIRMATIONS"], config["TRANSACTION_TIMEOUT"]
+        wallet = Wallet(web3, private_key, n_confirm, timeout)
 
-    print(f"alice_wallet.address = '{alice_wallet.address}'")
-    print(f"bob_wallet.address = '{bob_wallet.address}'")
+        print(f"{name}_wallet.address = '{wallet.address}'")
+        wallets.append(wallet)
 
-    return (alice_wallet, bob_wallet)
+    return tuple(wallets)
 
 
 def _remote_config(tmp_path):

